Fredzqm/Throttler return promise of handler result

fredzqm opened the pull request:
Depends on #1042 

Instead of throttling each delete task, we can throttling each REST requests.
However, we need to add the ability to obtain handler's result from the `Queue` or `Stack`.

I added `throttle<R>(task: T): Promise<R>`, which returns a promise with the task result.

`chunkedDelete` is now short and concise. We make "Promise tree" to keep track of deletion progress rather than tracking them manually in the queue.
Note that the FILO behavior is still essential here to prevent memory issue.

Ideas credit to @yuchenshi.

fredzqm commented on src/deploy/hosting/uploader.js:
```diff
@@ -179,7 +179,7 @@ class Uploader {
   queuePopulate() {
     const pop = this.populateBatch;
-    this.populateQueue.add(pop, "batch" + (this.populateQueue.stats().total + 1));
+    this.populateQueue.add(pop);
```
I traced back. This was added in https://github.com/firebase/firebase-tools/commit/484e784777cb2ae770358444fa20292c13e5c4e6#diff-2041a1adf696cab777ab5e6e3670a9e9R176.

`Queue.add()` only takes one argument, and we were ignoring the second argument the whole time.
I discovered this bug because hosting test won't finish. https://travis-ci.org/firebase/firebase-tools/jobs/463079863.
Cheers to whoever added the integration test! It is actually catching some bugs.

fredzqm commented on src/throttler/throttler.ts:
```diff
@@ -91,20 +96,39 @@ export abstract class Throttler<T> {
+  /**
+   * Add the task to the throttler and return a promise of handler's result.
+   */
+  public throttle<R>(task: T): Promise<R> {
```
I am open for a better way of doing this.
Right now, the caller will define a type, and it will cast it.

The result type may vary for a single `Throttler` instance, but is the same for each specific call of `throttle`.

fredzqm commented:
@bkendall Ready for review.
A high level summary:
I am trying achieve the effect of https://www.npmjs.com/package/promise-throttle.

bkendall commented on src/database/remove.ts:
```diff
@@ -34,66 +33,29 @@ export default class DatabaseRemove {
-        }
-      });
+  private async chunkedDelete(path: string): Promise<void> {
+    switch (await this.jobStack.throttle<NodeSize>(() => this.remote.prefetchTest(path))) {
```
I'm all for one-liners, but the `await` within a `switch` is pushing it too far. Make the `await` its own line.

bkendall commented on src/database/remove.ts:
```diff
@@ -34,66 +33,29 @@ export default class DatabaseRemove {
+      case NodeSize.EMPTY:
+        return;
+      default:
+        throw new FirebaseError("Unexpected prefetch test result: " + test, { exit: 3 });
```
1) what is `test`?
2) use template strings: `Unexpected... ${test}`

bkendall commented on src/test/throttler/throttler.spec.ts:
```diff
@@ -225,6 +225,19 @@ const throttlerTest = (throttlerConstructor: ThrottlerConstructor) => {
+
+  it("should return the result of task", () => {
+    const handler = (task: number) => {
+      return Promise.resolve("result: " + task);
```
prefer template strings

bkendall commented on src/throttler/throttler.ts:
```diff
@@ -31,6 +31,12 @@ export interface ThrottlerStats {
+interface TaskData<T> {
+  task: T;
+  retryCount: number;
+  wait: { resolve: (result: any) => void; reject: (err: Error) => void } | undefined;
```
```typescript
wait?: { resolve: (result: any) => void; reject: (err: Error) => void };
```

optional parameters can be defined with `?` rather than `| undefined`

bkendall commented on src/throttler/throttler.ts:
```diff
@@ -41,14 +47,13 @@ export abstract class Throttler<T> {
   public retried: number = 0;
   public total: number = 0;
-  public tasks: { [index: number]: T } = {};
+  public taskDatas: { [index: number]: TaskData<T> } = {};
```
I'm not thrilled about `taskDatas` - maybe `taskDataMap`?

Also, while we're at it: we're in ES6! Use [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)! We shouldn't be using `{}` as much as we are...

bkendall commented on src/throttler/throttler.ts:
```diff
@@ -91,20 +96,39 @@ export abstract class Throttler<T> {
+   * When the task is completed, resolve will be called with handler's result.
+   * If this task fails after retries, reject will be called with the error.
+   */
+  public add(
```
FWIW, `public` is the default: https://www.typescriptlang.org/docs/handbook/classes.html

I'm thinking we should drop the `public` keyword - what do you think? We don't have a lint rule enforcing this... yet...

bkendall replied:
I think that makes sense too. Let me take a look and I'll get back to you.

bkendall commented on src/throttler/throttler.ts:
```diff
@@ -138,16 +163,17 @@ export abstract class Throttler<T> {
+          taskData.retryCount++;
           this.retried++;
-          await _backoff(this.retryCounts[cursorIndex], this.backoff);
+          await _backoff(taskData.retryCount, this.backoff);
```
I'm sorry I didn't catch this before:

Don't prepend private function names/values with `_`. They're already marked private, so the compiler will take care of that for us.

yuchenshi commented on src/throttler/throttler.ts:
```diff
@@ -91,20 +96,39 @@ export abstract class Throttler<T> {
+  public add(
+    task: T,
+    wait?: { resolve: (result: any) => void; reject: (err: Error) => void }
+  ): void {
```
I'd vote for having `add` return promises for all cases, so we don't need two separate methods and separate code paths. If the caller does not need the result, they can feel free to discard it, just like calling any other function that returns a promise. A promise is not expensive to create and a promise without `.then` handlers will not cause any memory leaks. That's just my personal opinion though and you're welcome to keep it as it as.

fredzqm replied:
I agree with @yuchenshi 
@bkendall Thoughts on this.

bkendall replied:
Why does the user of this function care about the `wait` argument? Basically, that's the thing that's exposing that it's implemented using Promises. If I `.add` something to a queue, I'd very much expect that to be synchronous. The thing it _does_ may be async, but adding to a queue should not require callbacks of any kind.

Not handling the promise _is_ a dangerous idea, as well. If it failed, you'd have an uncaught exception floating around, which will cause issues as well.

The public API for this should make sense for the user: they want to add a `task`. Does the _user_ care about the ability to get the results? I think that should be available to the user, but I _don't_ think this is the appropriate place to attach and get that information - it feels like `wait` is an implementation detail and that your first instinct is more right, that there should be a helper that wraps the task being added with whatever the Throttler needs, but does so synchronously.

fredzqm replied:
Sure, I will keep private `addHelper` to hide the wait argument.

yuchenshi replied:
Did you mean having `add(task) -> void` (without `wait` argument) and `throttle(task) -> Promise`? If so, then totally agree with it.

However, I find the `throttle` method name confusing here because `add` also does throttling. It should have a name that signifies that it adds the task, and then **returns a promise that is resolved when the task finishes**. The bold part makes it different from `add`. Something like `addAndWaitForResult(task)` (but shorter) would be better IMO.

fredzqm replied:
`run` would be a OK candidate here.

fredzqm commented:
Fixed the conflicts with the removing public PR.

fredzqm commented:
@bkendall 
I added handler result to type parameter as well.
I reverted the changes in `database/remove.ts` so this PR only for throttler to keep it small.

The changes I just reverted in `database/remove.ts` was more of a clean up, but doesn't really improve performance, but makes the style easier to read. That is not the final version I am aiming for.
I am also working on a query parameter to limit the delete size, so that prefetch is no longer necessary. That is going to be a huge performance improvement.
I will clean up `database/remove.ts` at that PR then.

bkendall commented on src/throttler/throttler.ts:
```diff
@@ -31,7 +31,13 @@ export interface ThrottlerStats {
+interface TaskData<T, R> {
+  task: T;
+  retryCount: number;
+  wait?: { resolve: (R: any) => void; reject: (err: Error) => void };
```
should this be:

```suggestion
  wait?: { resolve: (value: R) => void; reject: (err: Error) => void };
```

?

fredzqm replied:
Sorry, that was a typo.

bkendall commented on src/throttler/throttler.ts:
```diff
@@ -31,7 +31,13 @@ export interface ThrottlerStats {
+  wait?: { resolve: (R: any) => void; reject: (err: Error) => void };
+}
+
+export abstract class Throttler<T, R> {
```
last thing. Add a jsdoc block and describe what `T` and `R` are. It'll help people who haven't been picking it apart (like you and me) to understand how each is used.

fredzqm replied:
Sure thing.
